### DISCUSSION TITLE
New `figure.caption` isn't flexible in set rules

### DISCUSSION THREAD

**Discussion by Andrew15-5** [typst/typst#2133] [General]
[Finally](https://github.com/typst/typst/releases/tag/v0.8.0), I can now edit the caption separator (https://github.com/typst/typst/issues/2109#issuecomment-1714281537):

```typc
set figure.caption(separator: ". ")
```

But the problem is that [`figure.caption`](https://typst.app/docs/reference/meta/figure/#definitions-caption) only provides `position` and `separator` in the set rule, but, for example, the `supplement` is only obtainable in the show rule. So, In order to change the `supplement` for the `kind: raw` you have to redefine the whole `figure.caption` instead of changing it in the set rule like this:

```typc
show figure.where(kind: raw): set figure.caption(supplement: "Best code ever")
```

Note that this line have no effect (for some reason; maybe because there is no figure inside the figure):

```typc
show figure.where(kind: raw): set figure(supplement: "Best code ever")
```

There is another problem with using:

```typc
show figure.where(kind: raw): fig => {
  show figure.caption: cap => {
    ...
  }
  fig
}
```

Between `cap.supplement` and `cap.counter.display((cap or fig??).numbering)` there is a whitespace, that you have to make sure to include. The `set figure.caption(separator: ". ")` can override (includes) a whitespace between ordinal number and caption text, but the whitespace between supplement and ordinal number isn't configurable. And there is `numbering` in both `figure` and `figure.caption`, so which `numbering` is for which purposes? Are they identical? If so, then we can also pass the identical `supplement` from the `figure` to the `figure.caption`.

I will list thing that I want to be changed:
- add `supplement` which includes the whitespace (between supplement and ordinal number) in the `set figure.caption()`

I guess that's it... ^^'

**Comment by laurmaedje** [comment]
The supplement and numbering are owned by the `figure`. The fields are only mirrored on the caption for convenience. The ideal way would be `show figure.where(kind: raw): set figure(supplement: "Best code ever")` but that this doesn't work in this case is a separate (known) bug.

**Comment by laurmaedje** [comment]
The whole supplement thing probably needs to be reworked a bit alongside a counter/numbering rework...
